### PR TITLE
Clear yum cache before any yum call

### PR DESCRIPTION
**Pull request by r0x0d**
This change intends to call the `check_custom_repos_are_valid` function
as the first thing before any other checks. The reason for that is we
can guarantee that we will work with the most recent cache for
`yum`/`dnf`.

Jira reference (If any): [RHELC-251](https://issues.redhat.com/browse/RHELC-251)
Bugzilla reference (If any):

### CODE REVIEW

**Comment by r0x0d** [comment]
Looking through the code we don't seem to mess with anything related to repos or pkg manager before the checks start to take place. I think that by just moving the function at the top of the checks should do the trick. 

@SpyTec, do you think we should rewrite this function to not be a *check*, but rather, do this as a special operation of cleaning/making cache? Then this `check_custom_repos_are_valid` could be rewritten in another way to just check against custom repos, and not deal with cache generation.

**Review by SpyTec — APPROVED** [review]

**Comment by r0x0d** [comment]
> 



> > @SpyTec, do you think we should rewrite this function to not be a check, but rather, do this as a special operation of cleaning/making cache? Then this check_custom_repos_are_valid could be rewritten in another way to just check against custom repos, and not deal with cache generation.
> > Something like this (this is just an example): gist.github.com/r0x0d/f645a1f7b005ee96e9421954e27528db
> 
> @r0x0d That would be out of scope for the issue and can instead be tackled in [OAMG-5570](https://issues.redhat.com/browse/OAMG-5570). For this PR we should just move it to be in the beginning. We also need to do more verification to make sure using repolist will work for us and be faster

Gotcha! I will upload this gist to [OAMG-5570](https://issues.redhat.com/browse/OAMG-5570) in case anyone in the future wants to work on that and they can take a look at this prototype of implementation.

**Review by SpyTec — APPROVED** [review]

**Review by bocekm — CHANGES_REQUESTED** [review]
The main goal of this change should be making sure that the yum cache is up-to-date before we start doing some checks - especially the check whether the installed packages including kernel are up-to-date (`handled bycheck_package_updates()`). We can do that by clearing the yum cache. That's what we do already in the function for checking whether custom repos used by the customer are accessible: https://github.com/oamg/convert2rhel/blob/v0.25/convert2rhel/checks.py#L165.

However this cache cleaning happens only when custom repos are used (--no-rhsm option is used). Yet we need to clean the cache every time - because the `check_package_updates()` is called every time.

There's a disadvantage of performing the cache cleaning every time though - rebuilding the cache takes long (minutes). But we don't currently have a better/faster way of making sure the yum cash is up-to-date. We can try speeding it up in a subsequent PR.

**Comment by r0x0d** [comment]
@bocekm you think that this draft I made for the PR could solve the issue? https://gist.github.com/r0x0d/f645a1f7b005ee96e9421954e27528db

**Comment by bocekm** [comment]
> @bocekm you think that this draft I made for the PR could solve the issue? https://gist.github.com/r0x0d/f645a1f7b005ee96e9421954e27528db

Feel free to introduce the change from the gist to this PR, so that we can do the review using the github tools like code line comments and suggestions.
Based on what did you choose yum repolist as an alternative to yum makecache? I remember that repolist may not be what we're looking for https://github.com/oamg/convert2rhel/pull/196#discussion_r588298250. But maybe with clean cache it is the right command, hence my question.
Also, maybe the makecache is not necessary to run in the `generate_latest_yum_cache()`. What we can do is just clean the cache there and let the first next yum call build the cache with just the data it needs for the executed yum command. I have a feeling that makecache downloads way too many data, more than we need.

> We could also include a command (or environment var) like --skip-yum-cache if the user guarantees to us that the yum metadata/cache is new.

Yeah, it'd be useful to have for development purposes. But only for that. We can't get any guarantee on that from the use. For that reason I'd introduce env var like `CONVERT2RHEL_UNSUPPORTED_DONT_CLEAN_YUM_CACHE`. Not in this PR but in a separate one though. We may find out that not running makecache makes the subsequent yum calls faster.

Ccing @abadger.

File: convert2rhel/checks.py [hunk 1]
```diff
@@ -147,7 +147,8 @@ def check_readonly_mounts():
 
 
 def check_custom_repos_are_valid():
-    """To prevent failures past the PONR, make sure that the enabled custom repositories are valid.
+    """To prevent failures past the PONR, make sure that the enabled custom
+    repositories are valid.
```
Comment (bocekm) [inline convert2rhel/checks.py:151]: Why this change? The line was below 120 chars. Ideally, to follow [PEP 257](https://www.python.org/dev/peps/pep-0257/), the multiline docstring of functions start with a single summary line, followed by an empty line and then detailed description.

File: convert2rhel/main.py [hunk 2]
```diff
@@ -80,12 +80,17 @@ def main():
         loggerinst.task("Prepare: Show Red Hat software EULA")
         show_eula()
 
+        # clean yum cache metadata
```
Comment (bocekm) [inline convert2rhel/main.py:83]:
```suggestion
```
The comment adds no additional value over what's right below that.

File: convert2rhel/pkghandler.py [hunk 3]
```diff
@@ -933,3 +933,15 @@ def has_duplicate_repos_across_disablerepo_enablerepo_options():
             message += "\n%s" % repo
         message += "\nThis ambiguity may have unintended consequences."
         loggerinst.warning(message)
+
+
+def clean_yum_metadata():
+    """Remove cached metadata from yum"""
```
Comment (bocekm) [inline convert2rhel/pkghandler.py:939]:
```suggestion
    """Remove cached metadata from yum.
    
    This is to make sure that Convert2RHEL works with up-to-date data from repositories before, for instance, querying
    whether the system has the latest package versions installed, or before checking whether enabled repositories
    have accessible URLs.
    """
```

File: convert2rhel/pkghandler.py [hunk 4]
```diff
@@ -933,3 +933,15 @@ def has_duplicate_repos_across_disablerepo_enablerepo_options():
...
+    output, ret_code = call_yum_cmd(command="clean", args=["metadata"], print_output=False)
+    loggerinst.debug("Output of yum clean metadata %s" % output)
```
Comment (bocekm) [inline convert2rhel/pkghandler.py:941]:
```suggestion
    loggerinst.debug("Output of yum clean metadata:\n%s" % output)
```

File: convert2rhel/pkghandler.py [hunk 5]
```diff
@@ -933,3 +933,15 @@ def has_duplicate_repos_across_disablerepo_enablerepo_options():
...
+
+    if ret_code != 0:
+        loggerinst.warning("Failed to clean yum metadata %s" % output)
```
Comment (bocekm) [inline convert2rhel/pkghandler.py:944]:
```suggestion
        loggerinst.warning("Failed to clean yum metadata:\n%s" % output)
```

File: convert2rhel/pkghandler.py [hunk 6]
```diff
@@ -933,3 +933,15 @@ def has_duplicate_repos_across_disablerepo_enablerepo_options():
...
+        return
+
+    loggerinst.info("Yum cache metadata cleaned successfully.")
```
Comment (bocekm) [inline convert2rhel/pkghandler.py:947]:
```suggestion
    loggerinst.info("Cached yum metadata cleaned successfully.")
```

**Review by bocekm — CHANGES_REQUESTED** [review]

**Review by bocekm — CHANGES_REQUESTED** [review]

File: convert2rhel/unit_tests/pkghandler_test.py [hunk 7]
```diff
@@ -1660,3 +1660,12 @@ def test_find_pkg_names(output, message, expected_names):
 def test_find_pkg_names_no_names(output, message):
     """Test that find_pkg_names does not find any names in these outputs."""
     assert pkghandler.find_pkg_names(output, message) == set()
+
+
+@pytest.mark.parametrize(
+    ("ret_code", "expected"), ((0, "Yum cache metadata cleaned successfully."), (1, "Failed to clean yum metadata"))
```
[the remark this answers is not in the document]
Comment (r0x0d) [reply convert2rhel/unit_tests/pkghandler_test.py:1666]: Oops! Forgot that.

[the remark this answers is not in the document]
Comment (r0x0d) [reply convert2rhel/unit_tests/pkghandler_test.py]: Done.

**Review by bocekm — DISMISSED** [review]
Approved. Waiting on an integration test.

**Comment by r0x0d** [comment]
Fixing the conflicts

**Comment by danmyway** [comment]
Hi @r0x0d !
The tests are failing mainly on this:
```[03/28/2022 18:19:17] DEBUG - Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/convert2rhel/main.py", line 84, in main
    pkghandler.clean_yum_metadata()
  File "/usr/lib/python3.6/site-packages/convert2rhel/pkghandler.py", line 980, in clean_yum_metadata
    output, ret_code = call_yum_cmd(command="clean", args=["metadata"], print_output=False)
  File "/usr/lib/python3.6/site-packages/convert2rhel/pkghandler.py", line 151, in call_yum_cmd
    if system_info.version.major == 8:
AttributeError: 'NoneType' object has no attribute 'major'
```
Could you please look into it?

**Comment by r0x0d** [comment]
> Hi @r0x0d ! The tests are failing mainly on this:
> 
> ```
>   File "/usr/lib/python3.6/site-packages/convert2rhel/main.py", line 84, in main
>     pkghandler.clean_yum_metadata()
>   File "/usr/lib/python3.6/site-packages/convert2rhel/pkghandler.py", line 980, in clean_yum_metadata
>     output, ret_code = call_yum_cmd(command="clean", args=["metadata"], print_output=False)
>   File "/usr/lib/python3.6/site-packages/convert2rhel/pkghandler.py", line 151, in call_yum_cmd
>     if system_info.version.major == 8:
> AttributeError: 'NoneType' object has no attribute 'major'
> ```
> 
> Could you please look into it?

That's odd :thinking:. 

No worries, will take a look right now.

**Comment by danmyway** [comment]
All tests have passed except for those with custom repos checks, which is expected as they will need some re-work.

File: convert2rhel/main.py [hunk 8]
```diff
@@ -85,7 +85,11 @@ def main():
         systeminfo.system_info.resolve_system_info()
         breadcrumbs.breadcrumbs.collect_early_data()
 
+        loggerinst.task("Prepare: Clean yum cache metadata")
+        pkghandler.clean_yum_metadata()
+
         # check the system prior the conversion (possible inhibit)
+        loggerinst.task("Prepare: Perform basic system checks")
         checks.perform_pre_checks()
```
Comment (r0x0d) [inline convert2rhel/main.py:93]: Codecov is mad because there is no unit test for this specific function, should we try to implement a basic mock test of this function?

File: convert2rhel/main.py [hunk 9]
```diff
@@ -85,8 +85,12 @@ def main():
         systeminfo.system_info.resolve_system_info()
         breadcrumbs.breadcrumbs.collect_early_data()
 
+        loggerinst.task("Prepare: Clean yum cache metadata")
+        pkghandler.clean_yum_metadata()
+
         # check the system prior the conversion (possible inhibit)
-        checks.perform_pre_checks()
+        loggerinst.task("Prepare: Perform basic system checks")
+        print(checks.perform_pre_checks())
```
Comment (abadger) [inline convert2rhel/main.py:93]:
```suggestion
```

Looks like debugging that was left in by mistake :-)   removing it should fix the lgtm check.

Comment (r0x0d) [reply convert2rhel/main.py:93]: Whoopsie! will remove that asap

**Review by bocekm — APPROVED** [review]

**Comment by bocekm** [comment]
@r0x0d, please rebase.

**Comment by r0x0d** [comment]
@bocekm rebased!

**Comment by bocekm** [comment]
Let's merge #458 first because the [`convert_offline_system` tests started to fail](http://artifacts.osci.redhat.com/testing-farm/7f5e8929-dce2-415a-b4f7-e238549a188c/work-convert_offline_systemsD0NqC8/log.txt) due to the merged #356.
```
01:03:55             out: [05/07/2022 01:03:55] TASK - [Prepare: Checking if the packages are up-to-date] *****************
01:03:55             out: Repository copr:copr.fedorainfracloud.org:group_oamg:convert2rhel is listed more than once in the configuration
01:03:55             out: Repository copr:copr.fedorainfracloud.org:group_oamg:convert2rhel is listed more than once in the configuration
01:03:55             out: Errors during downloading metadata for repository 'copr:copr.fedorainfracloud.org:group_oamg:convert2rhel':
01:03:55             out:   - Curl error (7): Couldn't connect to server for https://download.copr.fedorainfracloud.org/results/@oamg/convert2rhel/epel-8-x86_64/repodata/repomd.xml [Failed to connect to download.copr.fedorainfracloud.org port 443: Connection refused]
01:03:55             out: Error: Failed to download metadata for repo 'copr:copr.fedorainfracloud.org:group_oamg:convert2rhel': Cannot download repomd.xml: Cannot download repodata/repomd.xml: All mirrors were tried
01:03:55             out: Errors during downloading metadata for repository 'baseos':
01:03:55             out:   - Curl error (7): Couldn't connect to server for https://vault.centos.org/centos/8/BaseOS/x86_64/os/repodata/repomd.xml [Failed to connect to vault.centos.org port 443: Connection refused]
01:03:55             out: [05/07/2022 01:03:55] DEBUG - Traceback (most recent call last):
01:03:55             out:   File "/usr/lib/python3.6/site-packages/dnf/repo.py", line 574, in load
01:03:55             out:     ret = self._repo.load()
01:03:55             out:   File "/usr/lib64/python3.6/site-packages/libdnf/repo.py", line 397, in load
01:03:55             out:     return _repo.Repo_load(self)
01:03:55             out: libdnf._error.Error: Failed to download metadata for repo 'baseos': Cannot download repomd.xml: Cannot download repodata/repomd.xml: All mirrors were tried
```

**Comment by bocekm** [comment]
Actually merging #458 as suggested above would not solve all test issues. https://github.com/oamg/convert2rhel/pull/491 should.

**Comment by r0x0d** [comment]
Rebased code

**Comment by bocekm** [comment]
The tests fail on:
```
File "/usr/lib/python2.7/site-packages/convert2rhel/pkghandler.py", line 1038, in _get_packages_to_update_yum
 packages = base.returnPkgLists(["updates"], None)
...
RepoError: Cannot find a valid baseurl for repo: base/7/x86_64
```
Since we updated this function in https://github.com/oamg/convert2rhel/pull/495, @r0x0d, please rebase the PR to see if it fixes the problem.

**Comment by kokesak** [comment]
Hmm, is the `--skip-yum-cache` option implemented?
The tests are hitting errors for the offline conversions:
```
12:03:12             out: [06/14/2022 12:03:12] TASK - [Prepare: Gather system information] *******************************
12:03:12             out: Name:                Oracle Linux Server
12:03:12             out: OS version:          8.6
12:03:12             out: [06/14/2022 12:03:12] DEBUG - Calling command 'uname -i'
12:03:12             out: Architecture:        x86_64
12:03:12             out: Config filename:     oracle-8-x86_64.cfg
12:03:12             out: Skipping the execution of 'rpm -Va'.
12:03:12             out: [06/14/2022 12:03:12] DEBUG - Calling command 'uname -r'
12:03:12             out: [06/14/2022 12:03:12] DEBUG - Booted kernel VRA (version, release, architecture): 4.18.0-372.9.1.el8.x86_64
12:03:12             out: Checking internet connectivity using host '8.8.8.8' and port '53'.
12:03:15             out: WARNING - Couldn't connect to the host '8.8.8.8'.
12:03:15             out: 
12:03:15             out: [06/14/2022 12:03:15] TASK - [Prepare: Clear YUM/DNF version locks] *****************************
12:03:15             out: Usage of YUM/DNF versionlock plugin not detected.
12:03:15             out: 
12:03:15             out: [06/14/2022 12:03:15] TASK - [Prepare: Clean yum cache metadata] ********************************
12:03:15             out: [06/14/2022 12:03:15] DEBUG - Calling command 'yum clean metadata --quiet'
12:03:16             out: [06/14/2022 12:03:16] DEBUG - Output of yum clean metadata:
12:03:16             out: 
12:03:16             out: Cached yum metadata cleaned successfully.
12:03:16             out: 
12:03:16             out: [06/14/2022 12:03:16] TASK - [Prepare: Perform basic system checks] *****************************
12:03:16             out: 
12:03:16             out: [06/14/2022 12:03:16] TASK - [Prepare: Checking if --enablerepo repositories are accessible] ****
12:03:16             out: Skipping the check of repositories due to the use of RHSM for the conversion.
12:03:16             out: 
12:03:16             out: [06/14/2022 12:03:16] TASK - [Prepare: Checking the firmware interface type (BIOS/UEFI)] ********
12:03:16             out: BIOS detected.
12:03:16             out: [06/14/2022 12:03:16] DEBUG - Calling command 'grep ( /proc/modules'
12:03:16             out: 
12:03:16             out: [06/14/2022 12:03:16] TASK - [Prepare: Checking /mnt and /sys are read-write] *******************
12:03:16             out: [06/14/2022 12:03:16] DEBUG - /sys mount point is not read-only.
12:03:16             out: Read-only /mnt or /sys mount points not detected.
12:03:16             out: 
12:03:16             out: [06/14/2022 12:03:16] TASK - [Prepare: Check kernel compatibility with RHEL] ********************
12:03:16             out: [06/14/2022 12:03:16] DEBUG - Booted kernel version '4.18.0' corresponds to the version available in RHEL 8
12:03:16             out: [06/14/2022 12:03:16] DEBUG - Calling command 'rpm -qf --qf %{NAME} /boot/vmlinuz-4.18.0-372.9.1.el8.x86_64'
12:03:16             out: [06/14/2022 12:03:16] DEBUG - Booted kernel package name: kernel-core
12:03:16             out: [06/14/2022 12:03:16] DEBUG - The booted kernel is signed by Oracle.
12:03:16             out: Kernel is compatible with RHEL
12:03:16             out: 
12:03:16             out: [06/14/2022 12:03:16] TASK - [Prepare: Checking if the packages are up-to-date] *****************
12:03:16             out: Errors during downloading metadata for repository 'baseos':
12:03:16             out:   - Curl error (7): Couldn't connect to server for https://vault.centos.org/centos/8/BaseOS/x86_64/os/repodata/repomd.xml [Failed to connect to vault.centos.org port 443: Connection refused]
12:03:16             out: WARNING - There was an error while checking whether the installed packages are up-to-date. Having updated system is an important prerequisite for a successful conversion. Consider stopping the conversion to verify that manually.
12:03:16             out: WARNING - Failed to download metadata for repo 'baseos': Cannot download repomd.xml: Cannot download repodata/repomd.xml: All mirrors were tried
```
I suppose running `yum clean` make the following check fail as the system forgets some information. It tries to create a new record about the repos, but the URL is obviously not reachable as the system is "offline"
I thought that when the check for the internet connectivity fails, the up to date check is skipped :D

**Comment by r0x0d** [comment]
@kokesak we skip the check for update packages without internet only when we detect the hardcoded repos in the that specific folder they live. 

One thing that I can do, which seems more appropriate, is skipping that metadata clean call when there is no internet connection available. 

But in the other hand, if the user has everything set-up in a way where they don't have internet access, but they can access the repo in some way, we could get mismatch results because of this.

**Review by danmyway — DISMISSED** [review]
LGTM! Thank you, @kokesak

File: convert2rhel/pkghandler.py [hunk 10]
```diff
@@ -1088,3 +1088,25 @@ def _get_packages_to_update_dnf(reposdir):
...
+
+    This is to make sure that Convert2RHEL works with up-to-date data from repositories before, for instance, querying
+    whether the system has the latest package versions installed, or before checking whether enabled repositories have
+    accessible URLs.
+    """
+    # We are using run_subprocess here as an alternative to call_yum_cmd
+    # which doesn't apply the correct --enablerepos option because at this point
+    # the tool didn't initialize the necessary functions in SystemInfo.
+    # The viable solution was calling the yum command as a subprocess manually
+    # instead of using that function wrapper.
```
Comment (SpyTec) [inline convert2rhel/pkghandler.py:1104]: Great comment!

File: convert2rhel/unit_tests/main_test.py [hunk 11]
```diff
@@ -324,3 +327,63 @@ def test_post_ponr_conversion(monkeypatch):
...
+    assert system_release_file_mock.call_count == 1
+    assert os_release_file_mock.call_count == 1
+    assert backup_yum_repos_mock.call_count == 1
+    assert clear_versionlock_mock.call_count == 1
+    assert pre_ponr_conversion_mock.call_count == 1
+    assert ask_to_continue_mock.call_count == 1
+    assert post_ponr_conversion_mock.call_count == 1
+    assert rpm_files_diff_mock.call_count == 1
+    assert remove_tmp_dir_mock.call_count == 1
+    assert restart_system_mock.call_count == 1
```
[the remark this answers is not in the document]
Comment (SpyTec) [reply convert2rhel/unit_tests/main_test.py:389]: @r0x0d if you can, please do create one about refactoring `main.py` tests

**Comment by r0x0d** [comment]
Rebased.

**Comment by r0x0d** [comment]
Fixed the unit_test for centos7

**Comment by kokesak** [comment]
The failing integration tests are caused by incorrect cleanup in one of our tests. The fix for that is introduced here: https://github.com/oamg/convert2rhel/pull/447/commits/41fbb29e3060310253ed34df243985e96bbf709d

**Review by bocekm — APPROVED** [review]
One little nitpick otherwise it works and looks good.

**Review by bocekm — DISMISSED** [review]

File: tests/integration/tier0/basic-sanity-checks/test_basic_sanity_checks.py [hunk 12]
```diff
@@ -113,3 +113,13 @@ def test_c2r_latest_older_unsupported_version(convert2rhel):
...
     del os.environ["CONVERT2RHEL_UNSUPPORTED_VERSION"]
+
+
+def test_clean_cache(convert2rhel):
+    # Test that the yum clean is done before any other check that c2r does
+    with convert2rhel("--no-rpm-va --debug") as c2r:
+        c2r.expect("Prepare: Clean yum cache metadata")
+        c2r.expect("Cached yum metadata cleaned successfully.")
+        c2r.expect("Prepare: Perform basic system checks")
+        c2r.expect("Continue with the system conversion?")
```
Comment (danmyway) [inline tests/integration/tier0/basic-sanity-checks/test_basic_sanity_checks.py:124]:
```suggestion
        c2r.expect("Cached yum metadata cleaned successfully.")
        c2r.expect("Continue with the system conversion?")
```

Needs to be removed as it was removed in the main.py

File: tests/integration/tier0/basic-sanity-checks/test_basic_sanity_checks.py [hunk 13]
```diff
@@ -113,3 +113,13 @@ def test_c2r_latest_older_unsupported_version(convert2rhel):
     # Clean up
     os.system(f"cp /tmp/__init__.py {PATH_TO_VERSION}")
     del os.environ["CONVERT2RHEL_UNSUPPORTED_VERSION"]
+
+
+def test_clean_cache(convert2rhel):
+    # Test that the yum clean is done before any other check that c2r does
+    with convert2rhel("--no-rpm-va --debug") as c2r:
+        c2r.expect("Prepare: Clean yum cache metadata")
+        c2r.expect("Cached yum metadata cleaned successfully.")
```
Comment (danmyway) [inline tests/integration/tier0/basic-sanity-checks/test_basic_sanity_checks.py:122]:
```suggestion
        assert c2r.expect("Prepare: Clean yum cache metadata") == 0
        assert c2r.expect("Cached yum metadata cleaned successfully.") == 0

```

I would add assertion, so it possibly fails the assertion, thus not waiting for expect timeout if the string is not present for some reason.
WDYT, @kokesak ?

**Review by danmyway — CHANGES_REQUESTED** [review]

**Comment by danmyway** [comment]
I have run basic_sanity_checks with fixes from https://github.com/oamg/convert2rhel/commit/41fbb29e3060310253ed34df243985e96bbf709d and https://github.com/oamg/convert2rhel/pull/398#discussion_r929818090
All of them have passed.
We can either wait for https://github.com/oamg/convert2rhel/pull/447 to be merged and re-run the CI with rebased fixes or merge this PR after https://github.com/oamg/convert2rhel/pull/398#discussion_r929818090 is addressed.

**Review by danmyway — APPROVED** [review]
LGTM! Thank you @kokesak
